firedrake/preconditioners/pmg.py

```diff
@@ -0,0 +1,300 @@
+        V = test.function_space()
+        ele = V.ufl_element()
+        elements = [ele]
+        while True:
+            try:
+                ele = self.coarsen_element(ele)
+            except ValueError:
+                break
+            elements.append(ele)
+
+        pdm = PETSc.DMShell().create(comm=pc.comm)
+        sf = odm.getPointSF()
+        section = odm.getDefaultSection()
+        attach_hooks(pdm, level=len(elements)-1, sf=sf, section=section)
+        # Now overwrite some routines on the DM
+        pdm.setRefine(None)
+        pdm.setCoarsen(self.coarsen)
+        pdm.setCreateInterpolation(self.create_interpolation)
+        set_function_space(pdm, get_function_space(odm))
+
+        parent = get_parent(odm)
+        add_hook(parent, setup=partial(push_parent, pdm, parent), teardown=partial(pop_parent, pdm, parent),
+                 call_setup=True)
+        add_hook(parent, setup=partial(push_appctx, pdm, ctx), teardown=partial(pop_appctx, pdm, ctx),
+                 call_setup=True)
+
+        ppc = PETSc.PC().create(comm=pc.comm)
+        ppc.setOptionsPrefix(pc.getOptionsPrefix() + "pmg_")
+        ppc.setType("mg")
+        ppc.setOperators(*pc.getOperators())
+        ppc.setDM(pdm)
+        ppc.incrementTabLevel(1, parent=pc)
+
+        # PETSc unfortunately requires us to make an ugly hack.
+        # We would like to use GMG for the coarse solve, at least
+        # sometimes. But PETSc will use this p-DM's getRefineLevels()
+        # instead of the getRefineLevels() of the MeshHierarchy to
+        # decide how many levels it should use for PCMG applied to
+        # the p-MG's coarse problem. So we need to set an option
+        # for the user, if they haven't already; I don't know any
+        # other way to get PETSc to know this at the right time.
+        opts = PETSc.Options(pc.getOptionsPrefix() + "pmg_")
+        if "mg_coarse_pc_mg_levels" not in opts:
+            opts["mg_coarse_pc_mg_levels"] = odm.getRefineLevel() + 1
+
+        ppc.setFromOptions()
+        ppc.setUp()
+        self.ppc = ppc
+
+    def apply(self, pc, x, y):
+        return self.ppc.apply(x, y)
+
+    def applyTranspose(self, pc, x, y):
+        return self.ppc.applyTranspose(x, y)
+
+    def update(self, pc):
+        pass
+
+    def coarsen(self, fdm, comm):
+        fctx = get_appctx(fdm)
+        test, trial = fctx.J.arguments()
+        fV = test.function_space()
+        fu = fctx._problem.u
+
+        cele = self.coarsen_element(fV.ufl_element())
+        cV = firedrake.FunctionSpace(fV.mesh(), cele)
+        cdm = cV.dm
+        cu = firedrake.Function(cV)
+
+        parent = get_parent(fdm)
+        add_hook(parent, setup=partial(push_parent, cdm, parent), teardown=partial(pop_parent, cdm, parent),
+                 call_setup=True)
+
+        mapper = ArgumentReplacer({test: firedrake.TestFunction(cV),
+                                   trial: firedrake.TrialFunction(cV)})
```

Comment:
I think UFL's replace can now handle this case, can you check (and then purge `ArgumentReplacer` if so)?

Reply:
Purged.